machine/src/trace.rs

```diff
@@ -1,6 +1,17 @@
-pub trait TraceGenerator<M, T> {
+pub trait TraceGenerator<M> {
+    type F;
+    type FE;
     const NUM_COLS: usize;
+    const NUM_PERM_COLS: usize;
 
-    /// Generate a trace for the chip given the provided machine.
-    fn generate_trace(&self, machine: &M) -> Vec<[T; Self::NUM_COLS]>;
+    /// Generate the main trace for the chip given the provided machine.
+    fn generate_trace(&self, machine: &M) -> Vec<[Self::F; Self::NUM_COLS]>;
```

Comment:
Sorry I didn't notice this earlier, but my plan was that the prover would take a `trace: RowMajorMatrix<F>` (like in plonky3's baby-stark, although it's unfinished).

Happy to change it after. I think we can then also remove `generic_const_exprs`, which we're trying to move off of in Plonky2/Plonky3, as the current implementation seems to have serious issues.